DirectXTex/DirectXTexSwizzle.cpp

```diff
@@ -0,0 +1,303 @@
+//-------------------------------------------------------------------------------------
+// DirectXTexImage.cpp
+//
+// DirectX Texture Library - Image container
+//
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+//
+// http://go.microsoft.com/fwlink/?LinkId=248926
+//-------------------------------------------------------------------------------------
+
+#include "DirectXTex.h"
+#include "DirectXTexP.h"
+
+using namespace DirectX;
+using namespace DirectX::Internal;
+
+namespace
+{
+#ifdef __AVX2__
+#define deposit_bits(v,m) _pdep_u32(v,m)
+#else
+    // N3864 - A constexpr bitwise operations library for C++
+    // https://github.com/fmatthew5876/stdcxx-bitops
+    uint32_t deposit_bits(uint32_t val, uint32_t mask)
+    {
+        uint32_t res = 0;
+        for (uint32_t bb = 1; mask != 0; bb += bb)
+        {
+            if (val & bb)
+            {
+                res |= mask & (-mask);
+            }
+            mask &= (mask - 1);
+        }
+        return res;
+    }
+#endif
+}
+
+
+_Use_decl_annotations_
+HRESULT DirectX::StandardSwizzle(const Image& srcImage, bool toSwizzle, ScratchImage& result) noexcept
+{
+    if (!IsValid(srcImage.format))
+        return E_INVALIDARG;
+
+    if (IsTypeless(srcImage.format) || IsPlanar(srcImage.format) || IsPalettized(srcImage.format))
+        return HRESULT_E_NOT_SUPPORTED;
+
+    HRESULT hr = result.Initialize2D(srcImage.format, srcImage.width, srcImage.height, 1, 1);
+    if (FAILED(hr))
+        return hr;
+
+    Image uncompressedSource;
+    Image uncompressedDestination;
+    ScratchImage wide;
+    ScratchImage dest;
+    if (IsCompressed(srcImage.format))
+    {
+        wide = ScratchImage();
+        HRESULT de = Decompress(srcImage, srcImage.format, wide);
```

Comment:
Since Block Compression is lossy, decompress/swizzle/compress is not a high-quality solution. You can just swizzle it directly treating each block as if it were an 8byte or 16byte pixel. Same algorithm.

Reply:
Duplicate of https://github.com/microsoft/DirectXTex/pull/515#discussion_r1765398888